tree/readspeed/src/ReadSpeed.cxx

```diff
@@ -0,0 +1,314 @@
+// Author: Enrico Guiraud, David Poulton 2022
+
+/*************************************************************************
+ * Copyright (C) 1995-2022, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#include "ReadSpeed.hxx"
+
+#include <ROOT/TSeq.hxx>
+#include <ROOT/TThreadExecutor.hxx>
+#include <ROOT/TTreeProcessorMT.hxx>  // for TTreeProcessorMT::GetTasksPerWorkerHint
+#include <ROOT/InternalTreeUtils.hxx> // for ROOT::Internal::TreeUtils::GetTopLevelBranchNames
+#include <TBranch.h>
+#include <TStopwatch.h>
+#include <TTree.h>
+
+#include <algorithm>
+#include <cassert>
+#include <cmath> // std::ceil
+#include <memory>
+#include <stdexcept>
+#include <set>
+#include <regex>
+
+using namespace ReadSpeed;
+
+std::vector<std::string> ReadSpeed::GetMatchingBranchNames(const std::string &fileName, const std::string &treeName,
+                                                           const std::vector<std::string> &regexes)
+{
+   TFile *f = TFile::Open(fileName.c_str(), "READ_WITHOUT_GLOBALREGISTRATION");
+   if (f == nullptr || f->IsZombie())
+      throw std::runtime_error("Could not open file '" + fileName + '\'');
+   std::unique_ptr<TTree> t(f->Get<TTree>(treeName.c_str()));
+   if (t == nullptr)
+      throw std::runtime_error("Could not retrieve tree '" + treeName + "' from file '" + fileName + '\'');
+
+   const auto unfilteredBranchNames = ROOT::Internal::TreeUtils::GetTopLevelBranchNames(*t);
+   std::set<std::string> usedRegexes;
+   std::vector<std::string> branchNames;
+
+   auto filterBranchName = [regexes, &usedRegexes](std::string bName) {
+      const auto matchBranch = [&usedRegexes, bName](std::string regex) {
+         std::regex branchRegex(regex);
+         bool match = std::regex_match(bName, branchRegex);
+
+         if (match)
+            usedRegexes.insert(regex);
+
+         return match;
+      };
+
+      const auto iterator = std::find_if(regexes.begin(), regexes.end(), matchBranch);
+      return iterator != regexes.end();
+   };
+   std::copy_if(unfilteredBranchNames.begin(), unfilteredBranchNames.end(), std::back_inserter(branchNames),
```

Comment:
as a minor optimization we can skip this if `regexes.size() == 1 && regexes[0] == ".*"`